File: runtime-go/internal/encoder/security/dns.go
package security

import (
	"encoding/json"

	"github.com/linxinhong/eventweave/runtime-go/internal/encoder"
	"github.com/linxinhong/eventweave/runtime-go/internal/event"
)

// DNSJSON encodes DNS events as normalized JSON.
type DNSJSON struct{}

func init() {
	encoder.Register("dns-json", DNSJSON{})
}

// Name returns the encoder name.
func (DNSJSON) Name() string { return "dns-json" }

// ContentType returns the MIME type of the encoded output.
func (DNSJSON) ContentType() string { return "application/json" }

// Encode formats the event as a normalized DNS JSON record.
func (DNSJSON) Encode(ev event.Event) ([]byte, error) {
	if err := require(ev, []string{"client_ip", "query", "qtype"}); err != nil {
		return nil, err
	}
	record := map[string]any{
		"timestamp": ev.EventTime.Format("2006-01-02T15:04:05.000000Z"),
		"client_ip": ev.Attributes["client_ip"],
		"query":     ev.Attributes["query"],
		"qtype":     ev.Attributes["qtype"],
	}
	for key, value := range ev.Attributes {
		if _, exists := record[key]; !exists {
			record[key] = value
		}
	}
	return json.Marshal(record)
}

File: packs/security/encoders/dns.py
"""DNS JSON encoder."""

from __future__ import annotations

from eventweave.core.event import Event
from eventweave.encoders import Encoder, EncodeResult, encoder
from packs.security.encoders._helpers import format_json


@encoder("dns-json", content_type="application/x-ndjson")
class DnsJsonEncoder(Encoder):
    """Encode events as DNS query/response JSON."""

    name = "dns-json"
    content_type = "application/x-ndjson"

    _required = ["client_ip", "query"]

    def encode(self, event: Event) -> EncodeResult:
        missing = [f for f in self._required if f not in event.attributes]
        if missing:
            return self._fail(f"missing required fields: {', '.join(missing)}")

        record: dict[str, object] = {
            "client_ip": event.attributes["client_ip"],
            "query": event.attributes["query"],
        }
        optional = ["qtype", "rcode", "answers", "resolver", "action"]
        for key in optional:
            if key in event.attributes:
                record[key] = event.attributes[key]
        return self._ok(format_json(event, record))
